usnistgov/oar-pdr-py
# DBIO: tweak search API

The is PR is in response to a review of the [feature/dbio-selected-ep branch](https://github.com/usnistgov/oar-pdr-py/tree/feature/dbio-selected-ep) which introduces advance query capabilities based around the MongoDB filter query syntax.  In particular, it recommends the following changes:
   * change function name, `select_constraint_records()` to `adv_select_records()` for clarity
   * change signature for `select_records()` to allow for arbitrary keyword-value constraints to allow for future simple shortcut queries (e.g. `author=Plante` or `keywords=forensics`).
   * limit the available MongoDB operators via `check_query_structure()`

Unit tests have been updated to adapt to the API changes.

Please test in two ways
  * ensure that the unit tests pass; in this branch:
      1. run `scripts/testall.pyunit`  (make sure MONGO_TESTDB_URL is _not_ set)
      2. run the `test_mongo.py` script:
          1. Start the midas server stand-alone: `scripts/midasserver -b -M`
          2. set the MONGO_TESTDB_URL environment variable:  `export MONGO_TESTDB_URL=mongodb://oarop:oarop@localhost:27017/midas`
          3. ensure your PYTHONPATH is properly set (`export PYTHONPATH=$PWD/dist/pdr/lib/python`) and dependencies are installed.
          4. run the test script:  `python python/tests/nistoar/midas/dbio/test_mongo.py`
  * test under oar-docker:
     1. Check out the `feature/dbio-mongo` branch of `oar-docker`
     2. Change into the `oar-docker/midas` directory
     3. Build all components:  `../scripts/localdeploy`
     4. Build all containers: `../scripts/oarctl local build`
     5. Launch application: `../scripts/oarctl local up`
     6. Use portal to send advance queries to the DBIO service.
